### DISCUSSION TITLE
Open Seq2Seq Beam Search combined with Nemo Forced Aligner

### DISCUSSION THREAD

**Discussion by GreatHalf** [NVIDIA-NeMo/NeMo#5874] [Q&A]
Hello,

the generation of timestamps has been asked multiple times and solved with external tools ,see [parlance](https://github.com/NVIDIA/NeMo/discussions/2577) or [MFA](https://github.com/NVIDIA/NeMo/discussions/2657).

Now that NeMo contains a Forced Aligner, I would like to combine Seq2Seq Beam Search and the forced aligner. I asked this [before](https://github.com/NVIDIA/NeMo/discussions/5219) and back then you had to fall back on offline_diar_with_asr_infer.py, avoiding Seg2Seq entirely.

One new solution would be to combine Seq2Seq with the Forced Aligner: run eval_beamsearch_ngram.py first, use the resulting text as base truth for the forced aligner align.py. This doubles the creation of probs, which seems redundant to me. The creation of probs is by far the most time intensive part of both tasks.

Looking at the new align.py I would like to do this:
1. Generate probs once, for example [in this line](https://github.com/NVIDIA/NeMo/blob/a63bf61092c6f1802a1569705829b4f44f4d3eae/tools/nemo_forced_aligner/utils/data_prep.py#L303)
2. **To be developed:** Use the probs for a beam search as in eval_beamsearch_ngram.py, allowing you to use a kenlm_model_file, alpha and beta parameters. You can even use the same nemo CTC model. Save the resulting beam search transcript.
3. Use the beam search transcript as you would with align_using_pred_text.

Is this possible? Or are the probs tensors vastly different, e.g. by the size of the token alphabet? 
I see how the code would fit together, but I might be missing a theoretical problem.

Cheers

**Comment by GreatHalf** [comment]
I solved it (albeit the other way round for now). You can reuse the data probs pickle from  [eval_beamsearch_ngram.py](https://github.com/NVIDIA/NeMo/blob/main/scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram.py#L274) and skip the transcribe call in [align.py](https://github.com/NVIDIA/NeMo/blob/main/tools/nemo_forced_aligner/utils/data_prep.py#L303) entirely. Not much change is needed.


> import pandas
> …
>hypotheses = pickle.load("pickleForProbs.pickle")
>for hypothesis in hypotheses:
>>hypothesis = pandas.DataFrame(hypothesis)
>>hypothesis = torch.from_numpy(hypothesis.to_numpy())
>>log_probs_list_batch.append(hypothesis)
>>T_list_batch.append(hypothesis.shape[0])
>> pred_text_batch.append("irrelevant") # you could supply text from the *.tsv here. I suggest setting align_using_pred_text=FALSE and ignoring the pred_text_batch 
>>…

You might face the problem that by default the pickle file is for the whole manifest, while the manifest_lines_batch is just a part of the manifest.

If there is interest I will add a pickle file parameter to align.py and contribute.

Cheers

**Reply by titu1994** [reply]
@erastorgueva-nv to review

**Reply by GreatHalf** [reply]
PR is up [https://github.com/NVIDIA/NeMo/pull/9056](https://github.com/NVIDIA/NeMo/pull/9056)
